### PR TITLE
Make scanner's and api-resource-collector's limits configurable

### PR DESCRIPTION
**Pull request by jhrozek**
Even with the recent fix to the memory consumption, we might need to
provide a way for our users to bump the limits on scans, especially with
huge cluster nodes and Linux scans that traverse the filesystem.

This patch adds a new scanLimits attribute to the scan settings CR.
If unset, the compiled-in defaults are used, otherwise the settings take
precedence.

The only containers that are affected are the scanner and the
api-resource-collector as we've seen reports with those containers being
OOMKilled.

Jira: OCPBUGSM-39585

### CODE REVIEW

**Comment by jhrozek** [comment]
@mrogers950 @rhmdnd @Vincent056 because this PR touches the CRDs and tests, let me know if I should hold it before you merge the SDK upgrade.

**Comment by jhrozek** [comment]
@Wiharris do you think this enhancement could be documented as part of https://issues.redhat.com/browse/OSDOCS-4001 or would you prefer another ticket?

**Comment by Wiharris** [comment]
Hi Jakub,

I believe this enhancement could be documented with
https://issues.redhat.com/browse/OSDOCS-4001
<https://issues.redhat.com/browse/OSDOCS-4001> of another ticket; however,
I am good with whichever is easier for you. :)


Thanks,
William

**Comment by Wiharris** [comment]
lgtm! I have added the docs-approved label. :)

**Comment by jhrozek** [comment]
> lgtm! I have added the docs-approved label. :)

thanks for the fast review

**Comment by xiaojiey** [comment]
Detailed log seen from: https://privatebin.corp.redhat.com/?182e7cac751d75f2#DVRjoebGsZvxqCbM4RHhKcHGsSGpzTSzyUYuvftqn4Gx

**Comment by xiaojiey** [comment]
Detailed log seen from: https://privatebin.corp.redhat.com/?182e7cac751d75f2#DVRjoebGsZvxqCbM4RHhKcHGsSGpzTSzyUYuvftqn4Gx

**Comment by jhrozek** [comment]
/hold
in case @mrogers would like to merge the SDK bump first

**Comment by jhrozek** [comment]
rebased atop the new SDK

File: pkg/apis/compliance/v1alpha1/zz_generated.deepcopy.go
Comment (jhrozek) [inline pkg/apis/compliance/v1alpha1/zz_generated.deepcopy.go:300]: huh, wrong autogenerated code? That's a first for me..

Comment (jhrozek) [reply pkg/apis/compliance/v1alpha1/zz_generated.deepcopy.go:300]: @mrogers950 any idea? I don't know what I'm doing wrong

Comment (jhrozek) [reply pkg/apis/compliance/v1alpha1/zz_generated.deepcopy.go]: I decided to use a map in the CRD definition rather than the Resource type and now everything seems to work. From the user perspective this is the same, even for pods, the explain strings shows as a map:
```
oc explain pod.spec.containers.resources.limits
IND:     Pod                                                                  
VERSION:  v1                                                                   
                                                                               
FIELD:    limits <map[string]string>
...
```

File: config/manager/kustomization.yaml
Comment (mrogers950) [inline config/manager/kustomization.yaml:6]: One thing to note about config/manager/kustomization.yaml: The build targets run a kustomize "set image" command that will change this file in your tree, so it should not be committed. (make clean reverts it)

**Comment by jhrozek** [comment]
On Wed, Aug 31, 2022 at 08:05:41AM -0700, Matt Rogers wrote:
> @mrogers950 commented on this pull request.
> 
> 
> 
> > @@ -3,7 +3,7 @@ resources:
>  
>  images:
>  - name: compliance-operator
> -  newName: quay.io/compliance-operator/compliance-operator
> +  newName: image-registry.openshift-image-registry.svc:5000/openshift/compliance-operator
> 
> One thing to note about config/manager/kustomization.yaml: The build targets run a kustomize "set image" command that will change this file in your tree, so it should not be committed. (make clean reverts it)

Oops, fixed.

File: config/manager/kustomization.yaml
Comment (jhrozek) [reply config/manager/kustomization.yaml]: done

**Comment by jhrozek** [comment]
lol:
```
 error: error creating buildah builder: writing blob: adding layer with blob "sha256:56908b3674bda35f871058853518c895224107b74865114d1a5076b551728e2d": write /var/lib/containers/storage/overlay-layers/.tmp-b0d21825f59bef00ce7795b7dc6c37ac24090679b076103434b2cf0e81cc2f17.tar-split.gz3374980599: no space left on device 
```

File: tests/e2e/helpers.go
Comment (rhmdnd) [inline tests/e2e/helpers.go:9]: Shouldn't this be using `"github.com/ComplianceAsCode/compliance-operator`?

I suspect this is causing CI to fail:

```
 tests/e2e/helpers.go:9:2: cannot find package "." in:
	/go/src/github.com/ComplianceAsCode/compliance-operator/vendor/github.com/openshift/compliance-operator/pkg/controller/compliancescan
```

**Review by rhmdnd — CHANGES_REQUESTED** [review]
I think we have a typo that's causing the tests to fail.

File: config/crd/bases/compliance.openshift.io_compliancescans.yaml
Comment (rhmdnd) [inline config/crd/bases/compliance.openshift.io_compliancescans.yaml:202]: Do we, or should we, document defaults here?

File: CHANGELOG.md
Comment (rhmdnd) [inline CHANGELOG.md:21]: As an end user - when do I want to modify this attribute? What are the trade-offs of using higher and lower limits?

That might be a good thing to add here if we can distill it down to a sentence or two. Otherwise, we could add this to official documentation somewhere. Perhaps in the usage guide.

Comment (jhrozek) [reply CHANGELOG.md]: I added a sentence, does it read better?

File: config/crd/bases/compliance.openshift.io_compliancescans.yaml
Comment (jhrozek) [reply config/crd/bases/compliance.openshift.io_compliancescans.yaml:202]: I added a comment to the type itself, but for some reason it doesn't seem to propagate to the CRDs..ugh..

Comment (jhrozek) [reply config/crd/bases/compliance.openshift.io_compliancescans.yaml:202]: becuse it's `make manifests` /facepalm

File: CHANGELOG.md
Comment (rhmdnd) [inline CHANGELOG.md:23]: Yes - this is better. What's likely to cause the increased memory usage? The number of nodes I'm scanning or is the the number of rules in the profile, etc.?

Comment (mrogers950) [reply CHANGELOG.md:23]: We saw it primarily when users were able to deploy as watching all namespaces.

Comment (mrogers950) [reply CHANGELOG.md:23]: In which case it was a lot of namespace and that required watch cache for huge number of configmaps, deployments, etc.. If we have a lot of nodes, or rules, those are singular objects so the cache is not in as much danger.

Comment (jhrozek) [reply CHANGELOG.md:23]: Yes, I think also the nist profiles are quite memory intensive when scanning nodes because the number of RHCOS checks is high and some traverse the whole filesystem (or a large subset of).

Comment (jhrozek) [reply CHANGELOG.md:23]: btw when we asked the SCAP team what's the memory usage they see in the RHEL world their answer was IIRC "anywhere from a few hundred megs to ~2GB". They did some memory improvements since, but still..

Comment (rhmdnd) [reply CHANGELOG.md:23]: Ok - so if I deploy the operator to watch all namespaces, and I have a lot of namespaces, I'd be interested in this.

Comment (jhrozek) [reply CHANGELOG.md:23]: yes, the important thing being that if you install through the console (web ui) the default is to watch all namespaces.

**Review by rhmdnd — APPROVED** [review]
/lgtm

**Pull request merged**
